File: src/main/java/etlmail/front/gui/preferences/SavePrefences.java
package etlmail.front.gui.preferences;

import java.io.*;
import java.lang.reflect.InvocationTargetException;
import java.util.Properties;

import javax.annotation.PostConstruct;
import javax.swing.SwingUtilities;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationListener;
import org.springframework.stereotype.Component;

import com.google.common.io.Closeables;

import etlmail.front.gui.NewsletterNotificationBuilder;
import etlmail.front.gui.application.ShutdownEvent;

@Component
public class SavePrefences implements ApplicationListener<ShutdownEvent> {
    private static final String FILENAME = "mailgui.properties";

    private static final String TO = "to";
    private static final String FROM = "from";
    private static final String SUBJECT = "subject";
    private static final String TEMPLATE = "template";

    private static final String SERVER = "server";
    private static final String PORT = "port";
    private static final String USER = "user";

    private static final Logger log = LoggerFactory.getLogger(SavePrefences.class);

    private @Autowired NewsletterNotificationBuilder notificationBuilder;
    private @Autowired SwingServerConfiguration serverConfiguration;

    @PostConstruct
    public void init() {
	final Properties restored = readProperties();
	builderFromProperties(restored);
    }

    private Properties readProperties() {
	final Properties restored = new Properties();
	InputStream in = null;
	try {
	    in = new BufferedInputStream(new FileInputStream(FILENAME));
	    restored.load(in);
	    in.close();
	} catch (final IOException e) {
	    log.error("Cannot read preferences", e);
	} finally {
	    Closeables.closeQuietly(in);
	}
	return restored;
    }

    private void builderFromProperties(final Properties restored) {
	if (SwingUtilities.isEventDispatchThread()) {
	    restoreProperties(restored);
	} else {
	    SwingUtilities.invokeLater(new Runnable() {
		@Override
		public void run() {
		    restoreProperties(restored);
		}
	    });
	}
    }

    private void restoreProperties(final Properties restored) {
	notificationBuilder.to(restored.getProperty(TO));
	notificationBuilder.from(restored.getProperty(FROM));
	notificationBuilder.subject(restored.getProperty(SUBJECT));
	final String template = restored.getProperty(TEMPLATE);
	if (template != null) {
	    notificationBuilder.template(new File(template));
	}

	serverConfiguration.setHost(restored.getProperty(SERVER));
	final String port = restored.getProperty(PORT);
	if (null != port) {
	    serverConfiguration.setPort(Integer.parseInt(port));
	}
	serverConfiguration.setUsername(restored.getProperty(USER));
    }

    @Override
    public void onApplicationEvent(ShutdownEvent event) {
	final Properties toSave = propertiesFromBuilder();
	writeProperties(toSave);
    }

    private Properties propertiesFromBuilder() {
	final Properties toSave = new Properties();
	if (SwingUtilities.isEventDispatchThread()) {
	    setProperties(toSave);
	} else {
	    try {
		SwingUtilities.invokeAndWait(new Runnable() {
		    @Override
		    public void run() {
			setProperties(toSave);
		    }
		});
	    } catch (final InterruptedException e) {
		throw new IllegalStateException(e);
	    } catch (final InvocationTargetException e) {
		throw new IllegalStateException(e);
	    }
	}
	return toSave;
    }

    private void setProperties(final Properties toSave) {
	toSave.setProperty(TO, notificationBuilder.to());
	toSave.setProperty(FROM, notificationBuilder.from());
	toSave.setProperty(SUBJECT, notificationBuilder.subject());
	toSave.setProperty(TEMPLATE, notificationBuilder.template().getPath());

	toSave.setProperty(SERVER, serverConfiguration.getHost());
	toSave.setProperty(PORT, Integer.toString(serverConfiguration.getPort()));
	toSave.setProperty(USER, serverConfiguration.getUsername());
    }

    private void writeProperties(final Properties toSave) {
	OutputStream out = null;
	try {
	    out = new BufferedOutputStream(new FileOutputStream(FILENAME));
	    toSave.store(out, null);
	    out.close();
	} catch (final IOException e) {
	    log.error("Cannot save preferences", e);
	} finally {
	    Closeables.closeQuietly(out);
	}
    }
}

File: src/main/scala/etlmail/front/gui/preferences/SavePrefences.scala
package etlmail.front.gui.preferences

import java.io._
import java.util.Properties

import javax.annotation.PostConstruct

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationListener
import org.springframework.stereotype.Component

import com.google.common.io.Closeables

import com.weiglewilczek.slf4s.Logging

import etlmail.front.gui.application.InvokeAndWait
import etlmail.front.gui.application.ShutdownEvent
import etlmail.front.gui.sendmail.NewsletterNotificationBuilder

object SavePrefences {
  val FILENAME: String = "mailgui.properties"

  val TO: String = "to"
  val FROM: String = "from"
  val SUBJECT: String = "subject"
  val TEMPLATE: String = "template"

  val SERVER: String = "server"
  val PORT: String = "port"
  val USER: String = "user"
}

@Component
class SavePrefences extends ApplicationListener[ShutdownEvent] with Logging {
  import SavePrefences._

  @Autowired private var notificationBuilder: NewsletterNotificationBuilder = _
  @Autowired private var serverConfiguration: SwingServerConfiguration = _

  @PostConstruct
  def init() {
    fromProperties(readProperties())
  }

  private def readProperties(): Properties = {
    val restored = new Properties
    var in: InputStream = null
    try {
      in = new BufferedInputStream(new FileInputStream(FILENAME))
      restored.load(in)
      in.close()
    } catch {
      case e: IOException =>
        logger.error("Cannot read preferences", e)
    } finally {
      Closeables.closeQuietly(in)
    }
    return restored
  }

  @InvokeAndWait
  private def fromProperties(restored: Properties) {
    notificationBuilder.to(restored.getProperty(TO))
    notificationBuilder.from(restored.getProperty(FROM))
    notificationBuilder.subject(restored.getProperty(SUBJECT))
    val template = restored.getProperty(TEMPLATE)
    if (template != null) {
      notificationBuilder.template(new File(template))
    }

    serverConfiguration.setHost(restored.getProperty(SERVER))
    val port = restored.getProperty(PORT)
    if (null != port) {
      serverConfiguration.setPort(Integer.parseInt(port))
    }
    serverConfiguration.setUsername(restored.getProperty(USER))
  }

  @Override
  def onApplicationEvent(event: ShutdownEvent) {
    writeProperties(toProperties())
  }

  @InvokeAndWait
  private def toProperties(): Properties = {
    val toSave = new Properties
    toSave.setProperty(TO, notificationBuilder.to)
    toSave.setProperty(FROM, notificationBuilder.from)
    toSave.setProperty(SUBJECT, notificationBuilder.subject)
    toSave.setProperty(TEMPLATE, notificationBuilder.template.getPath)

    toSave.setProperty(SERVER, serverConfiguration.host)
    toSave.setProperty(PORT, Integer.toString(serverConfiguration.port))
    toSave.setProperty(USER, serverConfiguration.username)

    return toSave
  }

  private def writeProperties(toSave: Properties) {
    var out: OutputStream = null
    try {
      out = new BufferedOutputStream(new FileOutputStream(FILENAME))
      toSave.store(out, null)
      out.close()
    } catch {
      case e: IOException =>
        logger.error("Cannot save preferences", e)
    } finally {
      Closeables.closeQuietly(out)
    }
  }
}
